docs/proposals/config-management-plugin-v2.md

```diff
@@ -0,0 +1,304 @@
+which will pre-populate a volume shared between plugins and the repo-server.
+
+```bash
+# An init container will copy the argocd static binary into the shared volume
+# so that the CMP server can become the entrypoint
+initContainers:
+- command:
+  - cp
+  - -n
+  - /usr/local/bin/argocd
+  - /var/run/argocd/argocd-cmp-server
+  image: quay.io/argoproj/argocd:latest
+  name: copyutil
+  volumeMounts:
+  - mountPath: /var/run/argocd
+    name: var-files
+ 
+# var-files is a shared volume between repo-server and cmp-server which holds:
+# 1) socket files that repo-server uses to communicate to each plugin
+# 2) git repositories cloned by repo-server
+volumes:
+- emptyDir: {}
+  name: var-files
+```
+
+#### Configuration
+
+Plugins will be configured via a ConfigManagementPlugin manifest located inside the plugin container, placed at a 
+well-known location (e.g. /plugin.yaml). Argo CD is agnostic to the mechanism of how the plugin.yaml would be placed, 
+but various options can be used on how to place this file, including: 
+- Baking the file into the plugin image as part of docker build
+- Volume mapping the file through a configmap.
+
+Note that, while the ConfigManagementPlugin looks like a Kubernetes object, it is not actually a custom resource. 
+It only follows kubernetes-style spec conventions.
+
+```bash
+# metadata file is in the root and shell executor knows about it
+apiVersion: argoproj.io/v1alpha1
+kind: ConfigManagementPlugin
+metadata:
+  name: cdk8s
+spec:
+  version: v1.0
+  init:
+    command: [cdk8s, init]
+  generate:
+    command: [sh, -c, "cdk8s synth && cat dist/*.yaml"]
+  discovery:
+    find:
+    - command: [find . -name main.ts]
+      glob: "**/*/main.ts"
+    check:
+    - command: [-f ./main.ts]
+      glob: "main.ts"
+  allowConcurrency: true
+```
+
+#### Config Management Plugin API Server (cmp-server)
+The Config Management Plugin API Server (cmp-server) will be a new Argo CD component whose sole responsibility will be 
+to execute `generate` commands inside the plugin environment (the sidecar container), at the request of the repo-server.
+
+The cmp-server will expose the following APIs to the repo-server,
+
+- GenerateManifests(path) - returns YAML output using plugin tooling
+- IsSupported(path) - returns whether or not the given path is supported by the plugin
+
+At startup, cmp-server looks at the /plugin.yaml ConfigManagementPlugin specification file to understand how to perform the requests.
+
+#### Registration & Communication
+The repo-server needs to understand what all plugins are available to render manifests. To do this, the cmp-server 
+sidecars will register themselves as available plugins to the argocd-repo-server by populating named socket files in the 
+shared volume between repo-server and cmp-server. e.g.:
+
+```bash
+/var/run/argocd/plugins/
+                        cdk8s.sock
+                        jkcfg.sock
+                        pulumi.sock
```

Comment:
I would assume that two versions of plugin `foo` may not be registered ? We should mention this in `non-goals` in that case.

Reply:
@sbose78 I have added a separate versioning section here https://github.com/argoproj/argo-cd/pull/5927/files#diff-26036f97484ae7af9debd1ddb87dfabef8f7d88dfbfcf40369ca01a90f75fd9bR209